Supporting DataGridView (and it's many related classes)

ndykman opened the pull request:
Okay, this is a big merge, but all of it is in support of DataGridView as requested in issue #216. I have gone over it a few times, but peer review and comments are very much welcomed and needed. 

There are some complex contract code that matches the if/throw code in the source bases. It's unclear if some of these really could be checked by the static checker, but they are there to match the source. Others are fairly straightforward.

Also, of note, there are some cases in which some obvious seem missing, but I couldn't see that the code actually would guarantee or require them (for example, a Height property always being >= 0). I am open to adding these, but I'm already doing a lot and I'm trying to be very conservative. 

I'm sure there are things I missed here. I just appreciate your consideration. Hopefully, this is useful down the road.

Commit:
Initial checkin of modified files for DataGridView contract (and it's many related classes)

Commit:
Merge branch 'master' into issue216

This is needed to insure that the pull requests match the master branch upstream.

Commit:
Various bug fixes and formatting changes. Includes addition of missing information for certain classes.

Commit:
Adding missing/modified files from master before pull request.

Commit:
Restoring original System.Windows.Forms.TabControl.cs source from master.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.AccessibleObject.cs:
```diff
@@ -0,0 +1,231 @@
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:System.Windows.Forms.AccessibleObject"/> class.
+        /// </summary>
+        public AccessibleObject()
```
Is it required?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.AutoCompleteStringCollection.cs:
```diff
@@ -0,0 +1,236 @@
+        /// <param name="index">The position at which to insert the string.</param><param name="value">The string to insert.</param>
+        public void Insert(int index, string value)
+        {
+            Contract.Requires(index >= 0 && index <= this.Count);
```
Can we use the same esnure as for `Add` method?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.AutoCompleteStringCollection.cs:
```diff
@@ -0,0 +1,236 @@
+        public void RemoveAt(int index)
+        {
+            Contract.Requires(index >= 0 && index < this.Count);
+            Contract.Ensures(this.Count == Contract.OldValue(this.Count)-1);
```
spacing: shoudl be `Contract.OldValue(Count) - 1`.

and can we avoid using `this`?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.BaseCollection.cs:
```diff
@@ -0,0 +1,108 @@
+        public void CopyTo(Array ar, int index)
+        {
+            Contract.Requires(ar != null);
+            Contract.Requires(index >= 0);
```
It seems that in previous file one contract statement was used: `Contract.Requires(index >= 0 && index <= Count)`.

Should we use consistent approach/

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.BoundsSpecified.cs:
```diff
@@ -0,0 +1,35 @@
+        Location = Y | X,
+        Size = Height | Width,
+        All = Size | Location,
+        None = 0,
```
Maybe move `None` to the beginning of this enum declaration/

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.Cursor.cs:
```diff
@@ -0,0 +1,225 @@
+        // Initializes a new instance of the <see cref="T:System.Windows.Forms.Cursor"/> class from the specified file.
+        // </summary>
+        // <param name="fileName">The cursor file to load. </param>
+        // public Cursor(string fileName)
```
We can add requires for this constructor:

`Contract.Requires(!string.IsNullOrEmpty(fileName))`

Constructor itself does not have this check, but it calls FileStream that do have this check in the `Init` method.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataFormats.cs:
```diff
@@ -0,0 +1,203 @@
+    /// <summary>
+    /// Provides static, predefined <see cref="T:System.Windows.Forms.Clipboard"/> format names. Use them to identify the format of data that you store in an <see cref="T:System.Windows.Forms.IDataObject"/>.
+    /// </summary>
+    
```
Redundant blank line.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataFormats.cs:
```diff
@@ -0,0 +1,203 @@
+        // Specifies a Windows bitmap format. This static field is read-only.
+        // </summary>
+        // 
+
```
Formatting issue...

ndykman replied:
Agreed.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewAutoSizeColumnModeEventArgs.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <returns>
+        /// The <see cref="T:System.Windows.Forms.DataGridViewColumn"/> with the <see cref="P:System.Windows.Forms.DataGridViewColumn.AutoSizeMode"/> property that changed.
+        /// </returns>
+        public DataGridViewColumn Column {get;}
```
This is not a valid C# syntax (before c#6.0).

SergeyTeplyakov replied:
And formatting issue detected:)

ndykman replied:
Agreed.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewAutoSizeColumnsModeEventArgs.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <returns>
+        /// An array of <see cref="T:System.Windows.Forms.DataGridViewAutoSizeColumnMode"/> values representing the previous values of the column <see cref="P:System.Windows.Forms.DataGridViewColumn.AutoSizeMode"/> properties.
+        /// </returns>
+        public DataGridViewAutoSizeColumnMode[] PreviousModes { get; }
```
Invalid property declaration.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <returns>
+        /// The <see cref="T:System.Windows.Forms.DataGridViewColumn"/> that contains the cell, or null if the cell is not in a column.
+        /// </returns>
+        public DataGridViewColumn OwningColumn { get; }
```
Invalid property declaration.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <returns>
+        /// The <see cref="T:System.Windows.Forms.DataGridViewRow"/> that contains the cell, or null if the cell is not in a row.
+        /// </returns>
+        public DataGridViewRow OwningRow {get;}
```
Invalid operoperty declaration + formatting issue.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <param name="rowIndex">The index of the cell's parent row.</param><exception cref="T:System.ArgumentOutOfRangeException">The specified <paramref name="rowIndex"/> is less than 0 or greater than the number of rows in the control minus 1. </exception><exception cref="T:System.InvalidOperationException"><see cref="P:System.Windows.Forms.DataGridViewCell.ColumnIndex"/> is less than 0, ind ...
+        public Rectangle GetContentBounds(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
```
I don't think that this requires is correct. Here what I see in sources:

```
  public Rectangle GetContentBounds(int rowIndex)
        {
            if (this.DataGridView == null)
            {
                return Rectangle.Empty;
            }
```

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        public Rectangle GetContentBounds(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
```
And I can't see this arg check in this method as well...

In default implementation I see this:

```
protected virtual Rectangle GetContentBounds(Graphics graphics, DataGridViewCellStyle cellStyle, int rowIndex)
        {
            return Rectangle.Empty;
        }
```

So I don't think we can force this contract here.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <param name="rowIndex">The row index of the cell.</param><param name="context">A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewDataErrorContexts"/> values that specifies the data error context.</param><exception cref="T:System.ArgumentOutOfRangeException">The specified <paramref name="rowIndex"/> is less than 0 or greater than the number of rows in the control m ...
+        public object GetEditedFormattedValue(int rowIndex, DataGridViewDataErrorContexts context)
+        {
+            Contract.Requires(this.DataGridView != null);
```
The same is true here. I don't the this checks in the source code.

Here what I see:

```
        public object GetEditedFormattedValue(int rowIndex, DataGridViewDataErrorContexts context)
        {
            if (this.DataGridView == null)
            {
                return null;
            }
            DataGridViewCellStyle dataGridViewCellStyle = GetInheritedStyle(null, rowIndex, false /*includeColors*/);
            return GetEditedFormattedValue(GetValue(rowIndex), rowIndex, ref dataGridViewCellStyle, context);
        }
```

ndykman replied:
Agreed, it seems that there are some places in which a null DataGridView is allowed. Fix on the way.

ndykman committed:
Removal of this in front of properties. Fixes around properties not using {get;} default syntax that is C# 6.0 only. Also fixes to contracts around DataGridViewCell.cs in terms of checks around DataGridView. Some formatting fixes.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <param name="rowIndex">The row index of the current cell.</param><exception cref="T:System.ArgumentOutOfRangeException">The value of the <see cref="P:System.Windows.Forms.DataGridViewElement.DataGridView"/> property of the cell is not null and the specified <paramref name="rowIndex"/> is less than 0 or greater than the number of rows in the control minus 1. </exception><exception cref ...
+        public virtual ContextMenuStrip GetInheritedContextMenuStrip(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
```
Again, this is incorrect precondition. See http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,af0d302fa10e7897

Basically, the precondition is: 'if DataGridView is not null then rowIndex should be from [0, Rows.Count).

And if DataGridView is null those invariants should not be true.

We can express this using following construct:

`Contract.Requires(DataGridView != null || (rowIndex >= 0 && rowIndex < DataGridView.Rows.Count);`
`Contract.Requires(DataGridView != null || (ColumnIndex >= 0);`

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        public virtual ContextMenuStrip GetInheritedContextMenuStrip(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
```
This should be `rowIndex >= 0 *&&* rowIndex < DataGridView.Rows.Count`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
+            Contract.Requires(this.ColumnIndex < 0);
```
This should be `ColumnIndex >= 0`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <param name="rowIndex">The index of the row containing the cell.</param><exception cref="T:System.ArgumentException">The cell is not contained within a <see cref="T:System.Windows.Forms.DataGridView"/> control and <paramref name="rowIndex"/> is not -1.-or-<paramref name="rowIndex"/> is not the index of the row containing this cell.</exception><exception cref="T:System.ArgumentOutOfRan ...
+        public virtual DataGridViewElementStates GetInheritedState(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
```
This one is not correct as well (see http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,8bf37b550e753667).

Should be something like this:

`Contract.Requires(DataGridView == null || rowIndex == -1, "When Grid is null, index should be -1")`
`Contract.Requires(DataGridView != null || (rowIndex >= 0 && rowIndex < DataGridView.Rows.Count);`

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        public virtual DataGridViewCellStyle GetInheritedStyle(DataGridViewCellStyle inheritedCellStyle, int rowIndex,
+            bool includeColors)
+        {
+            Contract.Requires(this.DataGridView != null);
```
This is the only method for now that trully have this precondition! (see http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,48f5a95ad5483bc1, this method throws InvalidOperationException when DataGridView is null).

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+            bool includeColors)
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
```
Should be `&&`, but not `||`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
+            Contract.Requires(this.ColumnIndex < 0);
```
Should be `>= 0`!

SergeyTeplyakov commented:
I see a bunch of typos in contracts. I would suggest to give this PR a try and use this new version of contracts definition for your own project. It should be a very good sanity check. 

I would say that this should be required for such changes. I'll try to look at every precondition, but this stuff is REALLY time consuming...

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <param name="rowIndex">The index of the cell's parent row.</param><exception cref="T:System.InvalidOperationException"><paramref name="rowIndex"/> is -1</exception>
+        protected virtual Size GetSize(int rowIndex)
+        {
+            Contract.Requires(rowIndex != -1);
```
This precondition is also not correct. See http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,71451549454cc494

Should be:

`Contract.Requires(DataGridView == null || rowIndex != -1);`

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+        /// <param name="rowIndex">The index of the cell's parent row.</param><exception cref="T:System.ArgumentOutOfRangeException">The value of the <see cref="P:System.Windows.Forms.DataGridViewElement.DataGridView"/> property of the cell is not null and <paramref name="rowIndex"/> is less than 0 or greater than or equal to the number of rows in the parent <see cref="T:System.Windows.Forms.Data ...
+        protected virtual object GetValue(int rowIndex)
+        {
+            Contract.Requires(this.DataGridView != null);
```
OK, I'm giving up. The same chagnes should be made for all preconditions for other methods in this class...

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1265 @@
+    /// Represents an individual cell in a <see cref="T:System.Windows.Forms.DataGridView"/> control.
+    /// </summary>
+   
+    public abstract class DataGridViewCell : DataGridViewElement //, ICloneable, IDisposable
```
Almost all preconditions for this class are not correct.

Current implementation assumes that DataGridView property should not be null, but this is not the case.

I'm leaving this comment as a reminder for myself and for the author to double check all the preconditions for this class. I've made some proposals for some of them, but then I've realized that the same issue is happening over and over and I gave up.

ndykman replied:
Done.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs:
```diff
@@ -0,0 +1,164 @@
+        {
+            get
+            {
+                Contract.Requires(index >= 0 && index <= this.Count);
```
Of by one error. Should be `&& index < Count`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs:
```diff
@@ -0,0 +1,164 @@
+            }
+            set
+            {
+                Contract.Requires(value != null);
```
We can use stronger precondition (or add few more):

`Contract.Requires(value.DataGridView == null);`
`Contract.Requires(value.OwningRow == null);`

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs:
```diff
@@ -0,0 +1,164 @@
+            set
+            {
+                Contract.Requires(value != null);
+                Contract.Requires(value.DataGridView != null);
```
This one should be inverted:) As I mentioned in previous comment...

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs:
```diff
@@ -0,0 +1,164 @@
+        /// <param name="index">The zero-based index at which to place <paramref name="dataGridViewCell"/>.</param><param name="dataGridViewCell">The <see cref="T:System.Windows.Forms.DataGridViewCell"/> to insert.</param><exception cref="T:System.InvalidOperationException">The row that owns this <see cref="T:System.Windows.Forms.DataGridViewCellCollection"/> already belongs to a <see cref="T:Sys ...
+        public virtual void Insert(int index, DataGridViewCell dataGridViewCell)
+        {
+            Contract.Requires(index >= 0);
```
Should be `index >= 0 && index < Count`?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs:
```diff
@@ -0,0 +1,164 @@
+        public virtual void Insert(int index, DataGridViewCell dataGridViewCell)
+        {
+            Contract.Requires(index >= 0);
+        }
```
Two additional requirements could be added:

`Contract.Requires(dataGridViewCell != null);`
`Contract.Requires(dataGridViewCell.OwningRow == null);`

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs:
```diff
@@ -0,0 +1,164 @@
+        /// <param name="index">The zero-based index of the <see cref="T:System.Windows.Forms.DataGridViewCell"/> to be removed.</param><exception cref="T:System.InvalidOperationException">The row that owns this <see cref="T:System.Windows.Forms.DataGridViewCellCollection"/> already belongs to a <see cref="T:System.Windows.Forms.DataGridView"/> control.</exception><PermissionSet><IPermission clas ...
+        public virtual void RemoveAt(int index)
+        {
+            Contract.Requires(index >= 0);
```
This method should have the same precondition as indexer getter...

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellMouseEventArgs.cs:
```diff
@@ -0,0 +1,73 @@
+        public DataGridViewCellMouseEventArgs(int columnIndex, int rowIndex, int localX, int localY, MouseEventArgs e)
+          : base(e.Button, e.Clicks, localX, localY, e.Delta)
+        {
+
```
Redundant blank line.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellMouseEventArgs.cs:
```diff
@@ -0,0 +1,73 @@
+
+            Contract.Requires(columnIndex >= -1);
+            Contract.Requires(rowIndex >= -1);
+        }
```
I see another precondition here: `Contract.Requires(e != null)`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellPaintingEventArgs.cs:
```diff
@@ -0,0 +1,197 @@
+        /// <param name="clipBounds">A <see cref="T:System.Drawing.Rectangle"/> that specifies the area of the <see cref="T:System.Windows.Forms.DataGridView"/> to be painted.</param><param name="paintParts">A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewPaintParts"/> values specifying the parts to paint.</param><exception cref="T:System.InvalidOperationException"><see cre ...
+        public void Paint(Rectangle clipBounds, DataGridViewPaintParts paintParts)
+        {
+            Contract.Requires(!(this.RowIndex < -1));
```
Why to use so weird precondition? Why not to use `Requires(RowIndex >= -1)` as always?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellPaintingEventArgs.cs:
```diff
@@ -0,0 +1,197 @@
+        /// <param name="clipBounds">A <see cref="T:System.Drawing.Rectangle"/> that specifies the area of the <see cref="T:System.Windows.Forms.DataGridView"/> to be painted.</param><param name="cellsPaintSelectionBackground">true to paint the background of the specified bounds with the color of the <see cref="P:System.Windows.Forms.DataGridViewCellStyle.SelectionBackColor"/> property of the <se ...
+        public void PaintBackground(Rectangle clipBounds, bool cellsPaintSelectionBackground)
+        {
+            Contract.Requires(!(this.RowIndex < -1));
```
The same

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellPaintingEventArgs.cs:
```diff
@@ -0,0 +1,197 @@
+        /// <param name="clipBounds">A <see cref="T:System.Drawing.Rectangle"/> that specifies the area of the <see cref="T:System.Windows.Forms.DataGridView"/> to be painted.</param><exception cref="T:System.InvalidOperationException"><see cref="P:System.Windows.Forms.DataGridViewCellPaintingEventArgs.RowIndex"/> is less than -1 or greater than or equal to the number of rows in the <see cref="T: ...
+        public void PaintContent(Rectangle clipBounds)
+        {
+            Contract.Requires(!(this.RowIndex < -1));
```
The same

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellParsingEventArgs.cs:
```diff
@@ -0,0 +1,77 @@
+        /// The row index of the cell that was changed.
+        /// </returns>
+        /// 
+        public int RowIndex {get;}
```
Invalid (and not-formatted) C#.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellParsingEventArgs.cs:
```diff
@@ -0,0 +1,77 @@
+        /// The column index of the cell that was changed.
+        /// </returns>
+        /// 
+        public int ColumnIndex {get;}
```
Invalid and not-formatted C#

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellStyleContentChangedEventArgs.cs:
```diff
@@ -0,0 +1,43 @@
+        /// <returns>
+        /// The changed <see cref="T:System.Windows.Forms.DataGridViewCellStyle"/>.
+        /// </returns>
+        public DataGridViewCellStyle CellStyle { get; }
```
Invalid C#.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellToolTipTextNeededEventArgs.cs:
```diff
@@ -0,0 +1,42 @@
+        /// The current ToolTip text.
+        /// </returns>
+        // public string ToolTipText { get; set; }
+
```
Redundant blank line

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellValidatingEventArgs.cs:
```diff
@@ -0,0 +1,77 @@
+        public int ColumnIndex
+        {
+            get
+            {
```
I see some inconsistency in implementation for those args. Why this one has fields? but another one does not?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumn.cs:
```diff
@@ -0,0 +1,338 @@
+        /// <returns>
+        /// true if the column is connected to a data source; otherwise, false.
+        /// </returns>
+        public bool IsDataBound {get;}
```
Invalid and non-formatted C#.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumn.cs:
```diff
@@ -0,0 +1,338 @@
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<int>() >= 2);
```
Please introduce public constant and use it instead of using hardcoded value...

ndykman replied:
Where would the constant live? If I place it in the class, it will introduce a crasanitizer error (as it doesn't exist in the class).

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumn.cs:
```diff
@@ -0,0 +1,338 @@
+        public int Width
+        {
+            get { return default(int); }
+            set { Contract.Requires(value <= 65536); }
```
Please introduce a constant for this.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumn.cs:
```diff
@@ -0,0 +1,338 @@
+        public virtual int GetPreferredWidth(DataGridViewAutoSizeColumnMode autoSizeColumnMode, bool fixedHeight)
+        {
+            Contract.Requires(
+                !(autoSizeColumnMode == DataGridViewAutoSizeColumnMode.NotSet ||
```
Maybe != and && would be more readable?

ndykman replied:
Certainly, but for the more complex contracts, I've try to stick to just negating the condition that would cause a exception to be thrown. This is kind of the edge of that.

SergeyTeplyakov replied:
I meant to use following:

```c#
Contract.Requires(
                (autoSizeColumnMode != DataGridViewAutoSizeColumnMode.NotSet &&
autoSizeColumnMode != DataGridViewAutoSizeColumnMode.None &&
                  autoSizeColumnMode != DataGridViewAutoSizeColumnMode.Fill));
```

It seems a safe change but reads a lot better, IMO.

ndykman replied:
Agreed.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnCollection.cs:
```diff
@@ -0,0 +1,287 @@
+        {
+            get
+            {
+                Contract.Requires(index >= 0);
```
I think in most other cases we used one requires with two checks instead of one..

I.e. should be `index >= 0 && index < Count`.

ndykman replied:
Agreed. I will go through and merge these together when they occur.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnCollection.cs:
```diff
@@ -0,0 +1,287 @@
+        /// <param name="dataGridViewColumn">The <see cref="T:System.Windows.Forms.DataGridViewColumn"/> to add.</param><exception cref="T:System.ArgumentNullException"><paramref name="dataGridViewColumn"/> is null.</exception><exception cref="T:System.InvalidOperationException">The associated <see cref="T:System.Windows.Forms.DataGridView"/> control is performing one of the following actions tha ...
+        public virtual int Add(DataGridViewColumn dataGridViewColumn)
+        {
+            Contract.Requires(dataGridViewColumn != null);
```
What about postcondition? Is the result always greater than 0? Or >=0?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnCollection.cs:
```diff
@@ -0,0 +1,287 @@
+        /// <param name="array">The destination <see cref="T:System.Windows.Forms.DataGridViewColumn"/> array.</param><param name="index">The index of the destination array at which to start copying.</param>
+        public void CopyTo(DataGridViewColumn[] array, int index)
+        {
+            Contract.Requires(index >= 0);
```
Merge first two checks.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnCollection.cs:
```diff
@@ -0,0 +1,287 @@
+        {
+            Contract.Requires(index >= 0);
+            Contract.Requires(index < this.Count);
+            Contract.Requires(array != null);
```
Another precondition: `Contract.Requires(array.Rand == 1)`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnCollection.cs:
```diff
@@ -0,0 +1,287 @@
+        /// <param name="columnIndex">The zero-based index at which to insert the given column.</param><param name="dataGridViewColumn">The <see cref="T:System.Windows.Forms.DataGridViewColumn"/> to insert.</param><exception cref="T:System.ArgumentNullException"><paramref name="dataGridViewColumn"/> is null.</exception><exception cref="T:System.InvalidOperationException">The associated <see cref= ...
+        public virtual void Insert(int columnIndex, DataGridViewColumn dataGridViewColumn)
+        {
+            Contract.Requires(columnIndex >= 0);
```
&& columnIndex < Count.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnCollection.cs:
```diff
@@ -0,0 +1,287 @@
+        /// <param name="index">The index of the column to delete.</param><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="index"/> is less than zero or greater than the number of columns in the control minus one. </exception><exception cref="T:System.InvalidOperationException">The associated <see cref="T:System.Windows.Forms.DataGridView"/> control is performing one of the  ...
+        public virtual void RemoveAt(int index)
+        {
+            Contract.Requires(index >= 0);
```
Merge two statements.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnDividerDoubleClickEventArgs.cs:
```diff
@@ -0,0 +1,54 @@
+        /// </summary>
+        /// <param name="columnIndex">The index of the column next to the column divider that was double-clicked. </param><param name="e">A new <see cref="T:System.Windows.Forms.HandledMouseEventArgs"/> containing the inherited event data. </param><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="columnIndex"/> is less than -1.</exception>
+        public DataGridViewColumnDividerDoubleClickEventArgs(int columnIndex, HandledMouseEventArgs e)
+          : base(e.Button, e.Clicks, e.X, e.Y, e.Delta, e.Handled)
```
Add precondition: `Contract.Requires(e != null)`

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewColumnStateChangedEventArgs.cs:
```diff
@@ -0,0 +1,52 @@
+        /// The <see cref="T:System.Windows.Forms.DataGridViewColumn"/> whose state changed.
+        /// </returns>
+        /// 
+        public DataGridViewColumn Column {get;}
```
Invalid and non-formatted C# code.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewContentAlignment.cs:
```diff
@@ -0,0 +1,37 @@
+    public enum DataGridViewContentAlignment
+    {
+        NotSet = 0,
+        TopLeft = 1,
```
Maybe to use shifts? 1<<2, 1 << 3, 1 << 4?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewDataErrorContexts.cs:
```diff
@@ -0,0 +1,37 @@
+    /// </summary>
+    public enum DataGridViewDataErrorContexts
+    {
+        Formatting = 1,
```
The same question...

ndykman replied:
It might be a more readable for sure.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewDataErrorContexts.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// Represents the state of a data-bound <see cref="T:System.Windows.Forms.DataGridView"/> control when a data error occurred.
+    /// </summary>
+    public enum DataGridViewDataErrorContexts
```
Flags attribute is missing.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewContentAlignment.cs:
```diff
@@ -0,0 +1,37 @@
+    /// Defines constants that indicate the alignment of content within a <see cref="T:System.Windows.Forms.DataGridView"/> cell.
+    /// </summary>
+    
+    public enum DataGridViewContentAlignment
```
Flags attribute is missing.

ndykman replied:
It seems not to be defined on the type either from the disassembly. You would think it was from the values, but it isn't there as far as I can tell.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewContentAlignment.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// Defines constants that indicate the alignment of content within a <see cref="T:System.Windows.Forms.DataGridView"/> cell.
+    /// </summary>
+    
```
Redundant blank line.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewDataErrorEventArgs.cs:
```diff
@@ -0,0 +1,68 @@
+    
+    public class DataGridViewDataErrorEventArgs : DataGridViewCellCancelEventArgs
+    {
+
```
Redundant blank line

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewDataErrorEventArgs.cs:
```diff
@@ -0,0 +1,68 @@
+        public DataGridViewDataErrorEventArgs(Exception exception, int columnIndex, int rowIndex,
+            DataGridViewDataErrorContexts context)
+          : base(columnIndex, rowIndex)
+        {
```
Additional requires for columnIndex and rowIndex?

ndykman replied:
They are on the base constructor, should I add them here as well. If not, I will need to check for redundant contracts in subclasses.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewElement.cs:
```diff
@@ -0,0 +1,94 @@
+        /// A <see cref="T:System.Windows.Forms.DataGridView"/> control that contains this element. The default is null.
+        /// </returns>
+        /// 
+        public DataGridView DataGridView { get; }
```
INvalid C# code.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewHeaderCell.cs:
```diff
@@ -0,0 +1,249 @@
+{
+    public class DataGridViewHeaderCell : DataGridViewCell
+    {
+
```
Redundant blank line

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewHeaderCell.cs:
```diff
@@ -0,0 +1,249 @@
+            Contract.Requires(this.OwningRow == null ||
+                              !((this.DataGridView == null && rowIndex != -1 ||
+                                 this.DataGridView != null && (rowIndex < 0 || rowIndex >= this.DataGridView.Rows.Count))));
+            Contract.Requires(this.OwningColumn == null || this.DataGridView == null || rowIndex == -1);
```
I think it should be: `Requires(OwningRow == null || OwningColumn == null || rowIndex == -1)`. I can't see that DataGridView should be null.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewHeaderCell.cs:
```diff
@@ -0,0 +1,249 @@
+        protected override Size GetSize(int rowIndex)
+        {
+            Contract.Requires(this.OwningRow == null ||
+                              !((this.DataGridView == null && rowIndex != -1 ||
```
Because this is requires, this check should be inverted. You inverted only the first check (`OwningRow == null`). I don't think that this code is correct!

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewHeaderCell.cs:
```diff
@@ -0,0 +1,249 @@
+        /// <param name="rowIndex">The index of the cell's parent row.</param><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="rowIndex"/> is not -1.</exception>
+        protected override object GetValue(int rowIndex)
+        {
+            Contract.Requires(rowIndex == -1);
```
hm... This precondition looks super suspicios...:)

ndykman replied:
Indeed, but it is correct. Note the comment that an exception is raised if rowIndex is not -1. The disassembly confirms.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewPaintParts.cs:
```diff
@@ -0,0 +1,36 @@
+    public enum DataGridViewPaintParts
+    {
+        None = 0,
+        All = 127,
```
maybe move All at the bottom and use | to compute it?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+            get { return default(DataGridViewCellStyle); }
+            set
+            {
+                Contract.Requires(!(this.DataGridView != null && this.Index == -1));
```
Please use regular pattern for this: `Requires(DataGridView == null || Index != -1);`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+        {
+            get
+            {
+                Contract.Requires(!(this.DataGridView != null && this.Index == -1));
```
The same.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+            get { return default(int); }
+            set
+            {
+                Contract.Requires(!(this.DataGridView != null && this.Index == -1));
```
The same

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+        /// true if the row is frozen; otherwise, false.
+        /// </returns>
+        /// <exception cref="T:System.InvalidOperationException">The row is in a <see cref="T:System.Windows.Forms.DataGridView"/> control and is a shared row.</exception><PermissionSet><IPermission class="System.Security.Permissions.EnvironmentPermission, mscorlib, Version=2.0.3600.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" version="1" Unrestricted="true"/><IPermission class="System.Se ...
+        public bool Frozen
```
The same for both.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+        public int Height
+        {
+            get { return default(int); }
+            set { Contract.Requires(!(this.DataGridView != null && this.Index == -1)); }
```
The same

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+        public int MinimumHeight
+        {
+            get { return default(int); }
+            set { Contract.Requires(!(this.DataGridView != null && this.Index == -1)); }
```
The same

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+        {
+            get
+            {
+                Contract.Requires(!(this.DataGridView != null && this.Index == -1));
```
The same

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+                
+            }
+        }
+
```
The same issue with regular pattern for Index != -1 check for other properties.....

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+        public virtual int GetPreferredHeight(int rowIndex, DataGridViewAutoSizeRowMode autoSizeRowMode, bool fixedWidth)
+        {
+            // Contract.Requires(!((autoSizeRowMode & (DataGridViewAutoSizeRowMode) -4) != (DataGridViewAutoSizeRowMode)0));
+            Contract.Requires(!(this.DataGridView != null && (rowIndex < 0 || rowIndex >= this.DataGridView.Rows.Count)));
```
Please use consistent check:)

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+        [EditorBrowsable(EditorBrowsableState.Advanced)]
+        public virtual DataGridViewElementStates GetState(int rowIndex)
+        {
+            Contract.Requires(!(this.DataGridView != null && (rowIndex < 0 || rowIndex >= this.DataGridView.Rows.Count)));
```
The same

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRow.cs:
```diff
@@ -0,0 +1,710 @@
+            /// <exception cref="T:System.InvalidOperationException">The value of the <see cref="P:System.Windows.Forms.DataGridViewRow.DataGridViewRowAccessibleObject.Owner"/> property is null.</exception>
+            public int GetChildCount()
+            {
+                Contract.Requires(this.Owner != null);
```
Ensures result is >= 0?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        /// </returns>
+        // protected ArrayList List {get;}
+
+        internal ArrayList SharedList { get; }
```
Invalid C#

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        /// <returns>
+        /// The <see cref="T:System.Windows.Forms.DataGridView"/> that owns the <see cref="T:System.Windows.Forms.DataGridViewRowCollection"/>.
+        /// </returns>
+        protected DataGridView DataGridView {get;}
```
Invalid C#

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        // [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
+        public virtual int Add()
+        {
+            Contract.Requires(this.DataGridView.DataSource == null);
```
Requires DataGridView != null?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        public virtual int Add(params object[] values)
+        {
+            Contract.Requires(values != null);
+            Contract.Requires(this.DataGridView.VirtualMode == false);
```
Requires DataGridView != null?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        /// <param name="dataGridViewRow">The <see cref="T:System.Windows.Forms.DataGridViewRow"/> to add to the <see cref="T:System.Windows.Forms.DataGridViewRowCollection"/>.</param><exception cref="T:System.InvalidOperationException">The associated <see cref="T:System.Windows.Forms.DataGridView"/> control is performing one of the following actions that temporarily prevents new rows from being  ...
+        public virtual int Add(DataGridViewRow dataGridViewRow)
+        {
+            Contract.Requires(this.DataGridView.Columns.Count > 0);
```
DataGridView != null

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+            Contract.Requires(this.DataGridView.NoDimensionChangeAllowed == false);
+            Contract.Requires(this.DataGridView.RowTemplate.Cells.Count <= this.DataGridView.Columns.Count);
+            Contract.Ensures(Contract.Result<int>() >= 0);
+            Contract.Ensures(this.Count == Contract.OldValue(this.Count) + 1);
```
Maybe postcondition shoudl be `Cound == Contract.OldValue(Count) + count)`?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+            Contract.Requires(this.DataGridView.DataSource == null);
+            Contract.Requires(this.DataGridView.NoDimensionChangeAllowed == false);
+            Contract.Ensures(Contract.Result<int>() >= 0);
+            Contract.Ensures(this.Count == Contract.OldValue(this.Count) + 1);
```
I can't find a place that ensures that the size will grow.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+                !((includeFilter &
+                  ~(DataGridViewElementStates.Displayed | DataGridViewElementStates.Frozen |
+                    DataGridViewElementStates.ReadOnly | DataGridViewElementStates.Resizable |
+                    DataGridViewElementStates.Selected | DataGridViewElementStates.Visible)) !=
```
It should be easy to invert this check. Just remove first `!` and use `== None`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        public int GetFirstRow(DataGridViewElementStates includeFilter, DataGridViewElementStates excludeFilter)
+        {
+            Contract.Requires(
+                !((includeFilter &
```
The same as previous comment.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        {
+            Contract.Requires(indexStart < -1);
+            Contract.Requires(
+                !((includeFilter &
```
The same for everyone. But I think it would be useful to extract this stuff into method... Too verbose and error prone...

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+            Contract.Requires(this.DataGridView.NoDimensionChangeAllowed == false);
+            Contract.Requires(this.DataGridView.VirtualMode == false);
+            Contract.Requires(this.DataGridView.DataSource == null);
+            Contract.Requires(values.Length <= this.DataGridView.Columns.Count);
```
Ensures that Count == oldCount + values.Length?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        /// <param name="indexSource">The index of the row on which to base the new row.</param><param name="indexDestination">The position at which to insert the row.</param><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="indexSource"/> is less than zero or greater than the number of rows in the collection minus one.-or-<paramref name="indexDestination"/> is less than zero ...
+        public virtual void InsertCopy(int indexSource, int indexDestination)
+        {
+            this.InsertCopies(indexSource, indexDestination, 1);
```
Why no preconditions? I don't think that preconditions could be 'inherited' by calling another method.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs:
```diff
@@ -0,0 +1,614 @@
+        }
+
+        /// <summary>
+        /// Inserts rows into the collection at the specified po
```
Postcondition that count == oldCount - 1?

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.MouseButtons.cs:
```diff
@@ -0,0 +1,33 @@
+    [Flags]
+    public enum MouseButtons
+    {
+        Left = 1048576,
```
Maybe to use << operator?

SergeyTeplyakov replied:
Really hard to check that the value is correct.

ndykman replied:
Indeed. It's hard to see where these values came from.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.MouseEventArgs.cs:
```diff
@@ -25,52 +25,55 @@ namespace System.Windows.Forms {
+    ///
+    /// Returns:
+    ///     One of the System.Windows.Forms.MouseButtons values.
+    public MouseButtons Button { get; }
```
Invalid C# for all those properties...

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.ScrollableControl.cs:
```diff
@@ -0,0 +1,353 @@
+            get { return default(Size); }
+            set
+            {
+                Contract.Requires(value.Width >= 0 && value.Height >=0);
```
spacing issue. Should be `>= 0`.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.ScrollableControl.cs:
```diff
@@ -0,0 +1,353 @@
+        // </returns>
+        // <param name="bit">The flag to check.</param>
+        // protected bool GetScrollState(int bit)
+       
```
Redundant blank line.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.TextBoxBase.cs:
```diff
@@ -0,0 +1,709 @@
+        /// </returns>
+        protected virtual CreateParams CreateParams
+        {
+            [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.UnmanagedCode)]
```
Please remove this or comment out.

SergeyTeplyakov replied:
I meant `SecurityPermissionAttribute`.

ndykman commented:
I'll plow through this and get all the bugs out. Thanks for the help.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.ToolStripDropDown.cs:
```diff
@@ -0,0 +1,738 @@
+    /// </summary>
+    public class ToolStripDropDown : ToolStrip
+    {
+        
```
Redundant blank line

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.ToolStripDropDown.cs:
```diff
@@ -0,0 +1,738 @@
+        /// <returns>
+        /// The level of opacity for the form. The default is 1.00.
+        /// </returns>
+        [DefaultValue(1.0)]
```
Please remove [DefaultValue] attribute.

SergeyTeplyakov commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.ToolStripDropDownMenu.cs:
```diff
@@ -0,0 +1,128 @@
+    
+    public class ToolStripDropDownMenu : ToolStripDropDown
+    {
+
```
Redundant blank line.

SergeyTeplyakov commented:
I'm done for the review... Waiting for updated version.

panacekcz commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1272 @@
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<object>() == null);
```
Although this holds for the default implementation, subclasses can return other values, so this shouldn't be a postcondition.

panacekcz commented on Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs:
```diff
@@ -0,0 +1,1272 @@
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<System.Type>() == typeof(DataGridViewTextBoxEditingControl));
```
Subclasses should be allowed to return other types or null.

ndykman committed:
Significant fixes based on review. Includes formatting changes to reduce line counts and blank lines.
Includes numerous contract fixes and simplifications. Reduced the number of copy and paste errors.

ndykman committed:
Stray lines in ToolStripDropDown.cs fixed

ndykman committed:
Additional fixes around contracts in DataGridView.cs

ndykman committed:
Additional formatting fixes as well as simplification of contracts in DataGridViewRow.cs

ndykman commented:
I just finished adding a set of merges to this branch. I will go over one more time early next week to see if I can find any more obvious bugs. The fact I did this checking in VS2015 really bit me (C# 6.0 syntax for get only properties, etc.) and that added a lot of overhead. I still have a some questions on the use of constants, etc. as suggested here. Again,  DataGridView just has a lot of moving parts. 

Thanks again.

ndykman committed:
Added public const values for certain constants.

ndykman committed:
Merge branch 'master' into issue216

ndykman commented:
If you need anything from me in this regard, let me know. Sorry I've been out of touch, got caught up in other work.

SergeyTeplyakov commented:
I thought that not every comments were addressed. I'll double check my comments.

ndykman commented:
If you have any specific changes, just let me know. I think we can get this in after the next release.

SergeyTeplyakov commented:
Ok, last two questions:
1. Could you please confirm that build passes?
2. Is there any chances to run static checker on some reasonable sized projects with DataGridView to see the results? It could be your own sanity check on your own project. I'm not using static checker any more just because it's too annoying, but I'm warring that such kind of change can affect existing customers. Some sanity check would be useful (maybe you already did this and this stuff is not required at all).

P.S. I'll definitely merge the PR before next release of the tool! Thanks for your help.
P.S.S. Actually the first point is more important than second one:)

ndykman committed:
Merge branch 'master' into issue216. Adding merge to improve pull request
process and to check if the build passes.

ndykman commented:
Will do. I will check some basic projects when I get a chance. The build still works, and I will merge from the master. Also, again, this is not a critical path item in my mind, so if it needs to wait until after 1.10, that's fine by me.

ndykman commented:
It did pass a sanity check on a Outlook AddIn that was using some basic WIndows.Form windows. If anybody has a better project, just let me know. I am looking at try to find a better project for testing.

ndykman commented:
Just an update, still trying to find time to see if I can find a good code base to verify this is a reasonable set of contracts. But, until then, best not to merge something so big.

SergeyTeplyakov commented:
@ndykman Sounds good!
